### DISCUSSION TITLE
3.3 deprecations cause repeated TS types, `useStore`

### DISCUSSION THREAD

**Discussion by charkour** [pmndrs/zustand#1631] [Q&A]
With the recent changes and deprecations to the `create` function, I've found myself writing TS utilities repeatedly and I'm wondering if there is a better way to do this.

If I start with a vanilla store and want to use it frequently within React, I like to make a hook around it but the function signature is often reused. Is there a utility I'm missing?

```tsx
import { useStore } from 'zustand'

const useMyStore = <T,>(
  selector: (state: MyState) => T,
  equality?: (a: T, b: T) => boolean,
) => useStore(vanillaMyStore, selector, equality)
```

The old way of doing things was much shorter and cleaner. 

```tsx
import { create } from 'zustand'

const useMyStore = create(vanillaMyStore)
```

**Comment by dai-shi** [comment]
Thanks for the feedback. I wasn't aware how many people depend on that usage. I still think the deprecation is meaningful for minimal API surface.
I agree typing the bound store hook isn't easy.
But, in your case, you have a vanilla store already, so attaching api to a hook isn't necessary.
Is your concern about only typing in typescript?

**Reply by charkour** [reply]
Thanks Daishi

> But, in your case, you have a vanilla store already, so attaching api to a hook isn't necessary.

Can you elaborate on this statement? If the vanilla store is present, should I be calling `vanillaStore.getState()`?

My question actually stems from working on a good API for the [zundo middleware](https://github.com/charkour/zundo) for zustand. It's currently in v2 beta since I'm not completely happy with the API surface. Currently, zundo will make a vanilla store called `temporal` to track previous states and provide methods to navigate through that state. Most users of the library use it with React, so I'm not sure of the best way for them to convert the vanilla store to a reusable react hook. 

This is the [current documentation for zundo v2 beta](https://github.com/charkour/zundo) and the [main codesandbox example](https://codesandbox.io/s/currying-flower-2dom9?file=/src/App.tsx). 

Is this the expected way to create a reusable store hook from a vanilla store and type it properly? Thank you

**Reply by dai-shi** [reply]
> Can you elaborate on this statement? If the vanilla store is present, should I be calling `vanillaStore.getState()`?

Yeah, that's what I meant. Well, if you prefer `useFooState.getState()` style, nothing is stopping you, and you can type accordingly in your library.

> This is the [current documentation for zundo v2 beta](https://github.com/charkour/zundo)

Don't know... Seems a bit complicated, especially a new store being attached to a store.

```js
const useTemporalStore = create(useStore.temporal);
```
That looks a bit weird. Is it the deprecated usage?

**Reply by charkour** [reply]
Yeah, that's the deprecated usage. 

Attaching a store to another store is a little strange but this pattern works reasonably well. Would you approach this differently?

Thank you so much for your time

**Reply by dai-shi** [reply]
I don't know how it fits with zustand, but in valtio, `proxyWithHistory` attaches methods to the store itself, and only allows `value` prop for state value.

**Reply by charkour** [reply]
Awesome, thanks. I'll look at that for reference. 

v1 of zundo injected methods and state to the store itself, but felt that creating a second only for the temporal state would allow for more flexibility; however, I plan on having two middleware version, one that injects methods and one that creates another store.

**Reply by charkour** [reply]
I might make a helper funciton to convert a vanilla store to a react-based store to help the users of zundo.

**Reply by dai-shi** [reply]
Or, instead of making it a middleware, create an external util that takes an existing store and return something as you like.

**Reply by charkour** [reply]
Ah okay, so instead of making `temporal` be a middleware for a zustand store, have `temporal` be an external util that users can add to when they need state tracking through time for a store?

My reason to keep it as middleware was to make it easy for those in userland to add undo/redo to their stores and to make it chainable with other middleware. But I think there is some merit to switching from the middleware approach.

**Reply by charkour** [reply]
I'll consider that. Thank you and let me know if you have more ideas for this! 

This package is starting to grow in popularity so I'm hoping to be able to support the growing zustand ecosystem.

**Reply by dai-shi** [reply]
Personally, I generally don't like middleware very much, it's too hard to type things. Were you successful with typing `temporal` with the middleware?

**Reply by charkour** [reply]
So sorry for the late reply. I was able to type the `temproal` middleware. Here is the source: https://github.com/charkour/zundo/blob/main/packages/zundo/src/index.ts

I'm not opposed to changing from middleware to a utility function, but I'm thinking that change would come in a later version of zundo.
